pkg/trim-after-delete/main.go

```diff
@@ -0,0 +1,120 @@
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"os"
+	"os/exec"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/client"
+
+	"strings"
+	"sync"
+	"time"
+)
+
+// Listen for Docker image delete events and run a command after a delay.
+
+func main() {
+	// after-image-deletes --delay 10s -- /sbin/fstrim /var
+
+	delay := flag.Duration("delay", time.Second*10, "maximum time to wait after an image delete before triggering")
+	flag.Usage = func() {
+		fmt.Fprintf(os.Stderr, "%s: run a command after images are deleted by Docker.\n\n", os.Args[0])
+		fmt.Fprintf(os.Stderr, "Example usage:\n")
+		fmt.Fprintf(os.Stderr, "%s --delay 10s -- /sbin/fstrim /var\n", os.Args[0])
+		fmt.Fprintf(os.Stderr, "   -- run the command /sbin/fstrim /var at most 10s after an image is deleted.\n")
+		fmt.Fprintf(os.Stderr, "      This would allow large batches of image deletions to happen and amortise the\n")
+		fmt.Fprintf(os.Stderr, "      cost of the TRIM operation.\n\n")
+		fmt.Fprintf(os.Stderr, "Arguments:\n")
+		flag.PrintDefaults()
+	}
+
+	flag.Parse()
+	toRun := flag.Args()
+	if len(toRun) == 0 {
+		logrus.Errorf("Please supply a program to run")
+	}
+
+	logrus.Infof("I will run %s around %.1f seconds after an image is deleted", strings.Join(toRun, " "), delay.Seconds())
+
+	action := NewDelayedAction(*delay, func() {
+		logrus.Infof("Running %s", strings.Join(flag.Args(), " "))
+		cmd := exec.Command(toRun[0], toRun[1:]...)
+		err := cmd.Run()
+		if err != nil {
+			cmdline := strings.Join(toRun, " ")
+			if ee, ok := err.(*exec.ExitError); ok {
+				logrus.Warnf("Failed to run %s: %s", cmdline, string(ee.Stderr))
+				return
+			}
+			logrus.Warnf("Unexpected failure while to run %s: %#v", cmdline, err)
+		}
+	})
+
+RECONNECT:
+	for {
+		cli, err := client.NewEnvClient()
+		if err != nil {
+			logrus.Warnf("Failed to connect to the Docker daemon: will retry in 5s")
+			time.Sleep(time.Second)
+			continue
+		}
+		logrus.Infof("(Re-)connected to the Docker daemon")
+		events, errors := cli.Events(context.Background(), types.EventsOptions{})
+		for {
+			select {
+			case event := <-events:
+				if event.Action == "delete" && event.Type == "image" {
+					logrus.Infof("The delayed action will happen at least once more")
+					action.AtLeastOnceMore()
+				}
+			case err := <-errors:
+				logrus.Errorf("Caught error: %#v", err)
+				cli.Close()
+				time.Sleep(time.Second)
+				continue RECONNECT
+			}
+		}
+	}
+}
+
+// DelayedAction runs a function in the future at least once after every call
+// to AtLeastOnceMore.
+type DelayedAction struct {
+	d time.Duration // maximum time to wait before calling f
+	t *time.Timer   // if non-nil we can be sure that f is going to be called again
+	f func()        // the function to call in the future
+	m sync.Mutex
```

Comment:
This reminded me of http://www.jtolds.com/writing/2016/03/go-channels-are-bad-and-you-should-feel-bad/